src/spaces/utilities.jl

```diff
@@ -233,6 +235,158 @@ walk!(agent, ::typeof(rand), model::ABM{<:AbstractGridSpace{D}}; kwargs...) wher
 walk!(agent, ::typeof(rand), model::ABM{<:ContinuousSpace{D}}) where {D} =
     walk!(agent, Tuple(2.0 * rand(model.rng) - 1.0 for _ in 1:D), model)
 
+"""
+    randomwalk!(agent, model::ABM{<:AbstractGridSpace}, r; kwargs...)
+Move `agent` for a distance `r` in a random direction respecting boundary conditions
+and space metric.
+For Chebyshev and Manhattan metric, the step size `r` is rounded to `floor(Int,r)`;
+for Euclidean metric in a GridSpace, random walks are not defined due to a strong
+dependency on the given value of `r`.
+
+## Keywords
+- `ifempty` will check that the target position is unoccupied and only move if that's true.
+  By default this is true, set it to false if different agents can occupy the same position.
+  In a GridSpaceSingle, agents cannot overlap and this keyword has no effect.
+"""
+function randomwalk!(
+    agent::AbstractAgent,
+    model::ABM{<:AbstractGridSpace},
+    r::Real;
+    kwargs...
+)
+    if model.space.metric == :euclidean
+        throw(ArgumentError(
+            "Random walks on a GridSpace with Euclidean metric are not defined. " *
+            "You might want to use a ContinuousSpace or a different metric."
+        ))
+    end
+    offsets = offsets_at_radius(model, r)
+    walk!(agent, rand(offsets), model; kwargs...)
+end
+
+function randomwalk!(
+    agent::AbstractAgent,
+    model::ABM{<:GridSpaceSingle},
+    r::Real;
+    kwargs...
+) where D
+    if model.space.metric == :euclidean
+        throw(ArgumentError(
+            "Random walks on a GridSpace with Euclidean metric are not defined. " *
+            "You might want to use a ContinuousSpace or a different metric."
+        ))
+    end
+    offsets = offsets_at_radius(model, r)
+    target_positions = map(β -> normalize_position(agent.pos .+ β, model), offsets)
+    idx = findall(pos -> id_in_position(pos, model) == 0, target_positions)
+    available_offsets = @view offsets[idx]
+    if isempty(available_offsets)
+        # don't move, return agent only for type stability
+        agent
+    else
+        walk!(agent, rand(available_offsets), model; ifempty=false)
+    end
+end
+
+
+"""
+    rotate(w::SVector{2}, θ::Real)
+Rotate two-dimensional vector `w` by an angle `θ`.
+The angle must be given in radians.
+"""
+rotate(w::SVector{2}, θ::Real) = Angle2d(θ) * w
```

Comment:
In my experience it makes for more readable code if the functions are defined after they are called. So I'd put these `rotate` functions` _after_ the first `randomwalk!` function that actually calls then.